### PR TITLE
Sara and Selam Amperes Scrabble

### PR DESCRIPTION
**Pull request by SelamawitA**
# Scrabble
Congratulations! You're submitting your assignment.

## Comprehension Questions

Feature | Feedback
--- | ---
What are the advantages of using git when collaboratively working on one code base? |  The pair doesn't have to share one computer. Changes are clearly tracked, for easier reconciliation if there's any discrepancy. It's not necessary to manually manage versions of the file.  
Why do you think we chose to make the `score` method in the `Scoring` class a **class method** instead of an **instance method**? |  The larger scope of the class method is more appropriate for the Scrabble game because tracking the score transcends any particular instance. The values associated with the letters are constant throughout each instantiation, it is a path to ensuring all objects have access to the same data. 
Describe an example of a test you wrote for a _nominal case_. |  We created a test to make sure that if 4 tiles were drawn, there were 4 fewer tiles in the bag.
Describe an example of a test you wrote for an _edge case_. |  We created a test to make sure that if 100 tiles were drawn, the program would not return any tiles and instead returns an empty array. 
Does your code use any methods from the `Enumerable` mixin? If so, where and why was it helpful? | We used an Enumerable to count how many tiles were remaining in our tilebag. It was VERY helpful because it offered a concise way to add up the number of tiles in our bag. We also used an enumerable to traverse through the hash of letters in the TileBag class. It was helpful in reducing the amount of logic we were responsible for in accessing data within the data structure we built. 
What was one method you and your pair used to debug code? |  We used pry to help pick the best enumerable, awesome print with exit to see how our code was progressing, and puts()/print statements. 
Is there a specific piece of code you'd like feedback on? | The data structure that we used for the tilebag and the draw_tiles methods.
What are two discussion points that you and your pair discussed when giving/receiving feedback from each other that you would be willing to share? |  We talked about how there was a good balance of skills, and an vocal appreciation of the others' contributions. Their was also a good alignment in both curiosity and endurance, resulting in little friction while working through tricky issues.  |

### CODE REVIEW

File: lib/scoring.rb [hunk 1]
```diff
@@ -1,9 +1,63 @@
...
+      return total
     end
 
+    # Returns highest score from an array of words.
     def self.highest_score_from(array_of_words)
+      max = 0
+      if array_of_words.size == 0
+        highest_score = nil
+      elsif array_of_words.size == 1
+        highest_score = array_of_words[0]
```
Comment (CheezItMan) [inline lib/scoring.rb:44]: What about if the only word is invalid?

File: specs/scoring_spec.rb [hunk 2]
```diff
@@ -42,21 +42,39 @@
...
     end
 
     it 'returns the only word in a length-1 array' do
+      single_arr = ['word']
+      solo_arr_of_words = Scrabble::Scoring.highest_score_from(single_arr)
+      solo_arr_of_words.must_equal 'word'
     end
 
     it 'returns the highest word if there are two words' do
+      random_arr = ['zzebra','medium']
```
Comment (CheezItMan) [inline specs/scoring_spec.rb:57]: You should also vary up the order in these tests so that you can ensure that the order isn't determining the winner (unless it's supposed to).

```ruby
      random_arr = ['zzebra','medium']
      highest_of_words = Scrabble::Scoring.highest_score_from(random_arr)
      highest_of_words.must_equal 'zzebra'

      random_arr = ['medium','zzebra']
      highest_of_words = Scrabble::Scoring.highest_score_from(random_arr)
      highest_of_words.must_equal 'zzebra'
```

File: specs/player_spec.rb [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+      player_name = Scrabble::Player.new ("Ada")
+      must_return_word = player_name.play("lawl")
+      must_return_word.must_equal 7
+    end
+
+  end
+
+  describe 'total_score' do
+    it 'Returns a total score to the user' do
+      player_name = Scrabble::Player.new("Ada")
```
Comment (CheezItMan) [inline specs/player_spec.rb:80]: You should also test to verify that `total_sore` doesn't keep going up **after** the player wins.

File: specs/player_spec.rb [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+
+    it 'Returns false if player has already won' do
+      player_name = Scrabble::Player.new ("Ada")
+      player_name.play("ZZZZZZZ")
+      must_be_false = player_name.play("POP")
+      must_be_false.must_equal false
+    end
+
+    it 'Returns score of word if < 100 ' do
+      player_name = Scrabble::Player.new ("Ada")
```
Comment (CheezItMan) [inline specs/player_spec.rb:71]: What about playing invalid words?

File: specs/player_spec.rb [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+
+  describe 'won?' do
+    it 'Returns true if player has won' do
+      player_name = Scrabble::Player.new ("Ada")
+      player_name.play("ZZZZZZZ")
+      player_name.won?.must_equal true
+    end
+
+    it 'Returns false if player has not won' do
+      player_name = Scrabble::Player.new ("Ada")
```
Comment (CheezItMan) [inline specs/player_spec.rb:97]: What happens with `won?` if the player **never** played a word?

File: lib/player.rb [hunk 6]
```diff
@@ -0,0 +1,81 @@
+require 'awesome_print'
+MAX_NUM_OF_TILES_ALLOWED = 7
```
Comment (CheezItMan) [inline lib/player.rb:2]: Good use of a constant!

File: lib/player.rb [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+    # Returns array of players tiles.
+    def tiles
+      return @players_plaque
+    end
+
+    # Draws tiles from a TileBag to fill player's plaque.
+    def draw_tiles(tile_bag)
+      if @players_plaque.length < MAX_NUM_OF_TILES_ALLOWED
+        add_tiles = MAX_NUM_OF_TILES_ALLOWED  - @players_plaque.length
+        @players_plaque.concat(tile_bag.draw_tiles(add_tiles))
```
Comment (CheezItMan) [inline lib/player.rb:24]: Very good!

File: lib/player.rb [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+    def plays
+      return @array_of_words
+    end
+
+    # Adds the input word to the plays array
+    def play(word)
+      player_status = false
+      total_score = 0
+      @array_of_words.each do |word_in_arr|
+        total_score += Scrabble::Scoring.score(word_in_arr)
```
Comment (CheezItMan) [inline lib/player.rb:38]: You already have a `total_score` method inside `Player` why not use it?

File: lib/player.rb [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+      sum = 0
+      @array_of_words.each do |a_word|
+        sum+= Scrabble::Scoring.score(a_word)
+      end
+      return sum
+    end
+
+    # If the player has over 100 points, returns true, otherwise returns false
+    def won?
+      player_case = false
```
Comment (CheezItMan) [inline lib/player.rb:61]: This variable name really doesn't make sense.

File: lib/player.rb [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+    # Returns the highest scoring played word
+    def highest_scoring_word
+      highest_scoring_word = Scrabble::Scoring.highest_score_from(@array_of_words)
+
+      return highest_scoring_word
+    end
+
+    # Returns the highest_scoring_word score
+    def highest_word_score
+      return highest_scoring_value = Scrabble::Scoring.score(highest_scoring_word)
```
Comment (CheezItMan) [inline lib/player.rb:77]: **NIIIICE!**

File: lib/tilebag.rb [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+      # Returns a collection of random tiles, removes the tiles from the default set or returns an empty array and displays a message to the user.
+      def draw_tiles(num)
+        random_tiles = []
+        if num <= tiles_remaining && num > 1
+          num.times do
+            count = 0
+            random_number = rand(0...tiles_remaining)
+            @tile_bag.map{|letter_type,letter|
+              letter.each do |single_letter|
+                if count == random_number
```
Comment (CheezItMan) [inline lib/tilebag.rb:46]: The structure you've got here, means that each letter has an **equal** chance of being drawn.  Also after a letter is exhausted you'll continue to pop from the `letter` array.

**Comment by CheezItMan** [comment]
# Scrabble
## What We're Looking For

Feature | Feedback
--- | ---
**General** |
Answered comprehension questions | Check, well done
Both Teammates contributed to the codebase | Check both have commits!
Regular Commits with meaningful commit messages |  Good number of commits!!  Mostly great with commit messages, but avoid messages like `update player.rb`  Instead say what you did there.  
**Scoring class** |
`score` method |  Check
Uses appropriate data structure to store the letter score |  Your works, but is there a more concise way to store letter scores, maybe using a hash?
Appropriately handles edge cases |  Handles most edge cases, just missing an empty string.  
`highest_score_from` method |  One minor bug if the word chosen is invalid.  
Appropriately handles edge cases |  See above
Test implementations for `highest_score_from`|  Check, see my one note.
**Player class** |
Uses an array to keep track of words played |  Check
Uses existing code to create `highest_scoring_word` and `highest_word_score` |  Check
Returns `true` or `false` from `won?` method |  Check
Tests edge cases on the `play(word)` |  Not all, you don't test for invalid words.  
**TileBag class** |
Uses a data structure to keep the set of default tiles in the bag |  Your data structure has one problem, each letter has an **equal** chance to be drawn despite there being a lot more "E" tiles than "Q".  Also you could end up popping an empty array (see my notes).  
`draw_tiles` method uses the parameter to draw a set of random tiles |  Check
`tiles_remaining` method returns a count |  Check
Tests `draw_tiles` edge cases |  Check, well done, but you missed the subtle bug above.  
Summary | Overall nicely done.  You hit the learning goals with some minor bugs.  See my notes about testing and most especially about `TileBag`  Think about what different kind of structure you could have used.
